README.md
# Decode and encode text to binary #
### Node module and CLI to encode, decode and auto-detect binary numbers ### 
[![Build Status](https://travis-ci.com/TheChickenNagget/decode-encode-binary.svg?branch=master)](https://travis-ci.com/TheChickenNagget/decode-encode-binary)
![GitHub repo size](https://img.shields.io/github/repo-size/TheChickenNagget/decode-encode-binary?color=dgreen&label=Package%20Size&logo=github)
[![Total Downloads](https://badgen.net/npm/dt/decode-encode-binary)](https://www.npmjs.com/package/decode-encode-binary)
![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/decode-encode-binary)
[![Changelog](https://img.shields.io/badge/-Changelog%20%F0%9F%93%9D-blue)](https://github.com/TheChickenNagget/decode-encode-binary/blob/master/CHANGELOG.md)
![Line](https://github.com/TheChickenNagget/assets/raw/master/images/line.png) 
## Installation ##
##### Using [npm](https://www.npmjs.com/package/decode-encode-binary) #####
```shell
$ npm install decode-encode-binary
```
##### Using [Yarn](https://yarnpkg.com/en/package/decode-encode-binary) #####
```shell
$ yarn add decode-encode-binary
```
------
## Using in Terminal ##  
- The Terminal command is `encdec` or `decenc`, Both have same Functionallity
- Using in Terminal/Console Requires To Have package installed with `--global` Tag
```shell
$ encdec
$ > Enter your Input (Binary or Text)
$ > Hi
$ 0100100001101001
``` 
```shell 
$ decenc
$ > Enter your Input (Binary or Text)
$ > 0100100001101001
$ Hi
```
------
## Examples ##
#### An example **Decoding** text from binary numbers using decode function ####
```js
const test = require("decode-encode-binary")
console.log(test.decode("0100100001100101011011000110110001101111"))

// outputs "Hello"
```
#### An example **Encoding** text to binary numbers using encode function ####
```js
const test = require("decode-encode-binary")
console.log(test.encode("Hello"))

// outputs "0100100001100101011011000110110001101111"
 ```
#### An example using **Auto Detection** using auto function ####
```js
const test = require("decode-encode-binary")

console.log(test.auto("Hi"))
// outputs 0100100001101001

console.log(test.auto("0100100001101001"))  
// outputs "Hi"

```
------
#### Making space between every letter in binary ####
```js
const test = require("decode-encode-binary")
console.log(test.encode("Hi", true))

// outputs "01001000 01101001"

/*  Auto detection support */

console.log(test.auto("Hi", true))
// outputs "01001000 01101001"
```
[![Create New Issue](https://img.shields.io/badge/-Create%20Issue-red?style=for-the-badge)](https://github.com/TheChickenNagget/decode-encode-binary/issues/new)
[![Fork On GitHub](https://img.shields.io/badge/-Fork%20on%20Github-orange?style=for-the-badge)](https://github.com/TheChickenNagget/decode-encode-binary/fork)
[![Star On GitHub](https://img.shields.io/badge/-Star%20on%20Github-yellow?style=for-the-badge)](https://github.com/TheChickenNagget/decode-encode-binary/stargazers)
[![Try on Runkit](https://img.shields.io/badge/-Try%20on%20Runkit-green?style=for-the-badge)](https://npm.runkit.com/decode-encode-binary)
![Line](https://github.com/TheChickenNagget/assets/raw/master/images/line.png)


CHANGELOG.md
# 2.0.3
>   -    **Enhanced** Markdown Layout, Edited badges
>   -    **Removed** Functions.js Test file.
>   -    **Migrated** To New Version of [Travis CI](https://travis-ci.com/)
>   -    **Edited** The .npmignore file to exclude tests

# 2.0.1 - 2.0.2
>   -    **Improved** CLI
>   -    **Updated** Package.json & Readme file
>   -    **Updated** Travis.CI Configuration file
>   -    **Removed** Unused files from .github Folder

# 2.0.0
>   -    Commands were changed from `dec` & `enc` to `encdec` & `decenc`
>   -    **Heavily Updated** CLI, Migrated from manual args[0] to [inquirer.js](https://www.npmjs.com/package/inquirer)
>   -    **Uglified** Code to make it more user friendly
>   -    **Updated** old code in package.json
>   -    **Added** changelog.md


test.js
console.log(require("./index").auto("ayyo"))

test/tests.js
const fs = require("fs");
const colors = require('chalk');
const {execSync} = require("child_process")
const main = require("../src/index.js")

return fs.access(`${process.cwd()}/node_modules`, function(err) {
  if (err && err.code === 'ENOENT') {
  console.log("Installing Development Packages")
   execSync("npm install --only=dev",  {stdio: 'inherit'});
    return runTest();
   }else{ 
  runTest();
 }});


 function runTest(){
console.log(colors.red('-- Decode test --'));
function decodeResult() { return main.decode('01101000 01100101 01101100 01101100 01101111 00100000 01110111 01101111 01110010 01101100 01100100'); }
if (decodeResult() === 'hello world') {
  console.log(colors.green('[ SUCESS ]'));
  console.log(colors.yellow('------------'));
} else {
  console.log(colors.green('[ FAIL ]'));
  console.log(colors.yellow('------------'));
}
console.log(colors.red('-- Encode test --'));
function encodeResult() { return main.encode('hello world'); }
if (encodeResult() === '0110100001100101011011000110110001101111001000000111011101101111011100100110110001100100') {
  console.log(colors.green('[ SUCESS ]'));
  console.log(colors.yellow('------------'));
} else {
  console.log(colors.green('[ FAIL ]'));
  console.log(colors.yellow('------------'));
}
console.log(colors.red('-- Auto Encoding test --'));
function autoEResult() { return main.auto('hello world'); }
if (autoEResult() === '0110100001100101011011000110110001101111001000000111011101101111011100100110110001100100') {
  console.log(colors.green('[ SUCESS ]'));
  console.log(colors.yellow('------------'));
} else {
  console.log(colors.green('[ FAIL ]'));
  console.log(colors.yellow('------------'));
}

console.log(colors.red('-- Auto Decoding test --'));
function autoDResult() { return main.auto('0110100001100101011011000110110001101111001000000111011101101111011100100110110001100100'); }
if (autoDResult() === 'hello world') {
  console.log(colors.green('[ SUCESS ]'));
  console.log(colors.yellow('------------'));
} else {
  console.log(colors.green('[ FAIL ]'));
  console.log(colors.yellow('------------'));
}

console.log(colors.red('                              '));
console.log(colors.green('-- Running Advanced Tests.. --'));
console.log(colors.red('                              '));
console.log(colors.red('-- Auto Decoding with spaces --'));
function AutoEwSResult() { return main.auto('01101000 01100101 01101100 01101100 01101111 00100000 01110111 01101111 01110010 01101100 01100100'); }
if (AutoEwSResult() === 'hello world') {
  console.log(colors.green('[ SUCESS ]'));
  console.log(colors.yellow('------------'));
} else {
  console.log(colors.green('[ FAIL ]'));
  console.log(colors.yellow('------------'));
}

console.log(colors.red('-- Auto Decoding with spaces after every number --'));
function AutoEwSEResult() { return main.auto('0 1 1 0 1 0 0 0 0 1 1 0 0 1 0 1 0 1 1 0 1 1 0 0 0 1 1 0 1 1 0 0 0 1 1 0 1 1 1 1 0 0 1 0 0 0 0 0 0 1 1 1 0 1 1 1 0 1 10 1 1 1 1 0 1 1 1 0 0 1 0 0 1 10 1 1 0 0 0 1 1 0 0 1 0 0'); }
if (AutoEwSEResult() === 'hello world') {
  console.log(colors.green('[ SUCESS ]'));
  console.log(colors.yellow('------------'));
} else {
  console.log(colors.green('[ FAIL ]'));
  console.log(colors.yellow('------------'));
}}

test/testing.js
/*           
 ___       ___     ____     ______     ______        __      _    _____      __      _      _____   
(  (       )  )   (    )   (   __ \   (   __ \      /  \    / )  (_   _)    /  \    / )    / ___ \  
 \  \  _  /  /    / /\ \    ) (__) )   ) (__) )    / /\ \  / /     | |     / /\ \  / /    / /   \_) 
  \  \/ \/  /    ( (__) )  (    __/   (    __/     ) ) ) ) ) )     | |     ) ) ) ) ) )   ( (  ____  
   )   _   (      )    (    ) \ \  _   ) \ \  _   ( ( ( ( ( (      | |    ( ( ( ( ( (    ( ( (__  ) 
   \  ( )  /     /  /\  \  ( ( \ \_)) ( ( \ \_))  / /  \ \/ /     _| |__  / /  \ \/ /     \ \__/ /  
    \_/ \_/     /__(  )__\  )_) \__/   )_) \__/  (_/    \__/     /_____( (_/    \__/       \____/   
                                                                                                    
-- RUN "npm i --dev-only" BEFORE STARTING THIS --
- This is test file, its used for testing functions after some Function Update
- This is supposed to be used by developers but you can use it too if you wan't
- decodeTest - tests decode function with provided text
- encodeTest - tests encode function with provided text
- autoTest - tests auto-detect function with provided text
 ~ remove //'s when testing function ~
*/
const colors = require("chalk");
   console.log(colors.red("-- Decode test --"))
   function decodeResult(){ return require("../index.js").decode("01101000 01100101 01101100 01101100 01101111 00100000 01110111 01101111 01110010 01101100 01100100")}
   if(decodeResult() === "hello world"){ 
    console.log(colors.green(`[ SUCESS ]`)) 
    console.log(colors.yellow(`------------`))
   }else{
    console.log(colors.green(`[ FAIL ]`)) 
    console.log(colors.yellow(`------------`))
   }
   console.log(colors.red("-- Encode test --"))
  function encodeResult(){ return  require("../index.js").encode("hello world") }
    if(encodeResult() === "0110100001100101011011000110110001101111001000000111011101101111011100100110110001100100"){ 
    console.log(colors.green(`[ SUCESS ]`)) 
    console.log(colors.yellow(`------------`))
   }else{
    console.log(colors.green(`[ FAIL ]`)) 
    console.log(colors.yellow(`------------`))
   }
   console.log(colors.red("-- Auto Encoding test --"))
   function autoEResult(){ return  require("../index.js").auto("hello world") }
   if(autoEResult() === "0110100001100101011011000110110001101111001000000111011101101111011100100110110001100100"){ 
    console.log(colors.green(`[ SUCESS ]`)) 
    console.log(colors.yellow(`------------`))
   }else{
    console.log(colors.green(`[ FAIL ]`)) 
    console.log(colors.yellow(`------------`))
   }

   console.log(colors.red("-- Auto Decoding test --"))
   function autoDResult(){ return  require("../index.js").auto("0110100001100101011011000110110001101111001000000111011101101111011100100110110001100100") }
   if(autoDResult() === "hello world"){ 
    console.log(colors.green(`[ SUCESS ]`)) 
    console.log(colors.yellow(`------------`))
   }else{
    console.log(colors.green(`[ FAIL ]`)) 
    console.log(colors.yellow(`------------`))
   }

   console.log(colors.red("                              "))
   console.log(colors.green("-- Running Advanced Tests.. --"))
   console.log(colors.red("                              "))
   console.log(colors.red("-- Auto Decoding with spaces --"))
   function AutoEwSResult(){ return require("../index.js").auto("01101000 01100101 01101100 01101100 01101111 00100000 01110111 01101111 01110010 01101100 01100100")}
   if(AutoEwSResult() === "hello world"){
    console.log(colors.green(`[ SUCESS ]`)) 
    console.log(colors.yellow(`------------`))
   }else{
    console.log(colors.green(`[ FAIL ]`)) 
    console.log(colors.yellow(`------------`))
   }

   console.log(colors.red("-- Auto Decoding with spaces after every number (extreme) --"))
 function AutoEwSExtremeSResult(){ return require("../index.js").auto("0 1 1 0 1 0 0 0 0 1 1 0 0 1 0 1 0 1 1 0 1 1 0 0 0 1 1 0 1 1 0 0 0 1 1 0 1 1 1 1 0 0 1 0 0 0 0 0 0 1 1 1 0 1 1 1 0 1 10 1 1 1 1 0 1 1 1 0 0 1 0 0 1 10 1 1 0 0 0 1 1 0 0 1 0 0")}
 if(AutoEwSExtremeSResult() === "hello world"){  
    console.log(colors.green(`[ SUCESS ]`)) 
    console.log(colors.yellow(`------------`))
   }else{
    console.log(colors.green(`[ FAIL ]`)) 
    console.log(colors.yellow(`------------`))
   }



src/index.js
module.exports = {
  /**
     * Decodes a binary code to text
     * @param {string} binary
     * @example
     *
     * const test = require("decode-encode-binary")
     * console.log(test.decode("0100100001100101011011000110110001101111"))
     * // outputs "Hello"
     *
     */

  decode(b) {
    if (!b || /^\s*$/.test(b)) {
      throw new Error('No Decode Argument was Received');
    }
    if (typeof b === 'number') {
      throw new Error('Decode Argument must be string, received Number');
    }

    try {
      function tobin(b) {
        b = b.replace(/\s+/g, '');
        b = b.match(/.{1,8}/g).join(' ');
        return b
          .split(' ')
          .map((m) => String.fromCharCode(parseInt(m, 2)))
          .join('');
      }
      return tobin(b).toString();
    } catch (e) {
      if (e.message === "Cannot read property 'join' of null") {
        throw new Error("Decode argument can't be empty string");
      }
      throw new Error(e.stack);
    }
  },
  /**
     * Encodes a text to binary
     * @param {string} text
     * @param {boolean} spaces
     * @example
     * var test = require("decode-encode-binary")
     * console.log(test.encode("Hello"))
     * // outputs "0100100001100101011011000110110001101111"
     *
     * // second example using "spaces"
     * var test = require("decode-encode-binary")
     * console.log(test.encode("Hello", true))
     * // outputs "01001000 01100101 01101100 01101100 01101111"
     */
  encode(t, sso) {
    if (!t) {
      throw new Error('No Encode Argument was Received');
    }
    if (typeof t === 'number') {
      throw new Error('Encode Argument must be string, received Number');
    }
    try {
      function totxt(s, ss) {
        function zeroPad(n) {
          return '00000000'.slice(String(n).length) + n;
        }
        return t.replace(/[\s\S]/g, (t) => {
          t = zeroPad(t.charCodeAt().toString(2));
          if (sso === true) {
            return !1 === ss ? t : `${t} `;
          }
          return !1 === ss ? t : `${t}`;
        });
      }
      return totxt(t).toString();
    } catch (e) {
      throw new Error(`Error ${e.stack}`);
    }
  },

  /**
     * Automatically determine if you wan't to decode or encode
     * @param {string} detect
     * @example
     * var test = require("decode-encode-binary")
     * console.log(test.auto("Hello"))
     * // outputs "0100100001100101011011000110110001101111"
     *
     * console.log(test.auto("0100100001100101011011000110110001101111"))
     * // outputs "Hello"
     */

  auto(d, spcs) {
    if (/^[01][01\s]*[01]$/.test(d)) {
      return this.decode(d);
    }
    if (spcs === true) {
      return this.encode(d, true);
    }
    return this.encode(d);
  },
  
  version() {
    try {
      return require('../package.json').version;
    } catch (e) {
      throw new Error(e.stack);
    }
  },
};


src/browser.js
function encode(r,e){if(!r)throw new Error("No Text to encode was provided");if(" "===r)throw new Error("Text cannot be empty string");"number"==typeof r&&(r=r.toString());try{return(t=r,t.replace(/[\s\S]/g,r=>(r=function(r){return"00000000".slice(String(r).length)+r}(r.charCodeAt().toString(2)),!0===e?`${r} `:`${r}`))).toString()}catch(r){throw new Error(`Error ${r.stack}`)}var t}function decode(r){if(!r)throw new Error("No Text to decode was provided");if(" "===r)throw new Error("Text cannot be empty string");"number"==typeof r&&(r=r.toString());try{return(e=r,(e=(e=e.replace(/\s+/g,"")).match(/.{1,8}/g).join(" ")).split(" ").map(function(r){return String.fromCharCode(parseInt(r,2))}).join("")).toString()}catch(r){if("Cannot read property 'join' of null"===r.message)throw new Error("Text to decode cannot be empty string");throw new Error(r.stack)}var e}function auto(r,e){if(!r){if(!r)throw new Error("No Text to decode / encode was provided");if(" "===r)throw new Error("Text cannot be empty string")}if("number"==typeof r)throw new Error("please add ' ' tags to the binary numbers, for example: test.decode('000101101') instead of the: test.decode(000101101)");if(/^[01][01\s]*[01]$/.test(r))try{return(o=r,(o=(o=o.replace(/\s+/g,"")).match(/.{1,8}/g).join(" ")).split(" ").map(function(r){return String.fromCharCode(parseInt(r,2))}).join("")).toString()}catch(r){if("Cannot read property 'join' of null"===r.message)throw new Error("Text to decode cannot be empty string");throw new Error(r.stack)}else{if(!r)throw new Error("No Text to encode was provided");if(" "===r)throw new Error("Text cannot be empty string");if("number"==typeof r)throw new Error("Text to encode must be string instead of the number, for example use encode('4') instead of the encode(4)");try{return(t=r,t.replace(/[\s\S]/g,r=>(r=function(r){return"00000000".slice(String(r).length)+r}(r.charCodeAt().toString(2)),!0===e?`${r} `:`${r}`))).toString()}catch(r){throw new Error(`Error ${r.stack}`)}}var t,o}


src/cli-loader.js
// eslint-disable-next-line no-console
#!/usr/bin/env node

const inquirer = require("inquirer");
const main = require("./index.js");

inquirer
  .prompt([
    {
      name: "text",
      type: "input",
      prefix: "",
      message: "Enter Your Input (Binary or Text)",
    },
  ])
  .then((answers) => {
    // First check if string is binary number
    switch (/^[01][01\s]*[01]$/.test(answers.text)) {
      case true:
        console.log(main.decode(answers.text));
        break;
      case false:
        console.log(main.encode(answers.text));
        break;
    }
  })
  .catch((error) => {
    error = error.toString();

    if (error.isTtyError) {
      throw new Error("TTY Error happened");
    } else if (error.includes("No")) {
      console.error("! No string provided !");
    } else {
      throw new Error(error);
    }
  });


testing.js
/*           
 ___       ___     ____     ______     ______        __      _    _____      __      _      _____   
(  (       )  )   (    )   (   __ \   (   __ \      /  \    / )  (_   _)    /  \    / )    / ___ \  
 \  \  _  /  /    / /\ \    ) (__) )   ) (__) )    / /\ \  / /     | |     / /\ \  / /    / /   \_) 
  \  \/ \/  /    ( (__) )  (    __/   (    __/     ) ) ) ) ) )     | |     ) ) ) ) ) )   ( (  ____  
   )   _   (      )    (    ) \ \  _   ) \ \  _   ( ( ( ( ( (      | |    ( ( ( ( ( (    ( ( (__  ) 
   \  ( )  /     /  /\  \  ( ( \ \_)) ( ( \ \_))  / /  \ \/ /     _| |__  / /  \ \/ /     \ \__/ /  
    \_/ \_/     /__(  )__\  )_) \__/   )_) \__/  (_/    \__/     /_____( (_/    \__/       \____/   
                                                                                                    

- This is test file, its used for testing functions after some Function Update
- This is supposed to be used by developers but you can use it too if you wan't

- decodeTest - tests decode function with provided text
- encodeTest - tests encode function with provided text
- autoTest - tests auto-detect function with provided text

 ~ remove //'s when testing function ~
*/
function decodeTest(text){console.log(require("./index.js").decode(text))}
function encodeTest(text){console.log(require("./index.js").encode(text))}
function autoTest(text){console.log(require("./index.js").auto(text))}
    
   //decodeTest("01101000 01100101 01101100 01101100 01101111 00100000 01110111 01101111 01110010 01101100 01100100")
   //encodeTest("hello world")
   //autoTest("hello world")
   //autoTest("01101000 01100101 01101100 01101100 01101111 00100000 01110111 01101111 01110010 01101100 01100100")